enterprise/app/trends/heatmap.tsx

```diff
@@ -0,0 +1,422 @@
+import React from "react";
+import moment from "moment";
+
+import { stats } from "../../../proto/stats_ts_proto";
+import { ScaleBand, scaleBand } from "d3-scale";
+import { withResizeDetector } from "react-resize-detector";
+import { pinBottomLeftOffsetFromMouse, MouseCoords, Tooltip } from "../../../app/components/tooltip/tooltip";
+
+interface HeatmapProps {
+  heatmapData: stats.GetStatHeatmapResponse;
+  width: number;
+  height: number;
+  valueFormatter: (value: number) => string;
+  metricBucketName: string;
+  metricBucketFormatter: (value: number) => string;
+  selectionCallback?: (s?: HeatmapSelection) => void;
+}
+
+interface State {
+  selectionToRender?: [TooltipData, TooltipData];
+}
+
+type TooltipData = {
+  timestamp: number;
+  timestampBucketIndex: number;
+  metric: number;
+  metricBucketIndex: number;
+  value: number;
+};
+
+export type HeatmapSelection = {
+  dateRangeMicros: number[];
+  bucketRange: number[];
+  invocationsSelected: number;
+};
+
+const CHART_MARGINS = {
+  top: 1,
+  right: 1,
+  bottom: 39,
+  left: 89,
+};
+
+// This is a magic number that states there will only be one axis label for
+// every N pixels of rendered axis length (currently 100).
+const TICK_LABEL_SPACING_MAGIC_NUMBER = 100;
+const heatmapColorScale = [
```

Comment:
nit: is there a way to generate this color scale automatically? (if there is a tool you can link to here, that might be helpful, otherwise no worries)

Reply:
I wrote out what I did in words; if we wanted to check it in as a tool i'd be happy to do that, but it was just some dumb little for loop that i wrote.